worker/caasapplicationprovisioner/application.go

```diff
@@ -93,7 +99,19 @@ func (a *appWorker) Wait() error {
 	return a.catacomb.Wait()
 }
 
-func (a *appWorker) loop() error {
+func (a *appWorker) loop() (err error) {
+	defer func() {
+		select {
+		case <-a.catacomb.Dying():
+			// Destroying model.
+		default:
+			if err == nil {
+				// Stop and remove myself from runner.
+				a.shutDownCleanUpFunc()
```

Comment:
Normally we wouldn't need this here - we would use the parent worker to remove this app worker from the runner when the life went to Dead or the app is not found etc. Am I right in thinking this is done here because of the `return nil` below after calling `verifyCharmUpgraded`? Is it definitely the case that we need to do this? ie if we exit nil in the worker loop, do we not see the Dead or not found transition in the parent worker?

Reply:
Yes, instead of peacefully exiting, we need to remove it from the runner as well.
The life checking notfound handler in parent worker is only triggered by model teardown but not triggered by app removal.
Here is for app removal or verifyCharmUpgraded returns true.